R/cran-compliance.R

```diff
@@ -0,0 +1,221 @@
+
+  # set the path for the duration of the function
+  withr::local_dir(pkg_root)
+
+  if (is.null(lib_name)) {
+    lib_name <- as_valid_rust_name(pkg_name(path))
+  } else if (length(lib_name) > 1) {
+    cli::cli_abort(
+      "{.arg lib_name} must be a character scalar",
+      class = "rextendr_error"
+    )
+  }
+
+  # add configure and configure.win templates
+  use_rextendr_template(
+    "cran/configure",
+    save_as = "configure",
+    quiet = quiet,
+    overwrite = overwrite,
+    data = list(lib_name = lib_name)
+  )
+
+  # configure needs to be made executable
+  Sys.chmod("configure", "0755")
+
+  use_rextendr_template(
+    "cran/configure.win",
+    save_as = "configure.win",
+    quiet = quiet,
+    overwrite = overwrite,
+    data = list(lib_name = lib_name)
+  )
+
+  # use CRAN specific Makevars templates
+  use_rextendr_template(
+    "cran/Makevars",
+    save_as = file.path("src", "Makevars"),
+    quiet = quiet,
+    overwrite = overwrite,
+    data = list(lib_name = lib_name)
+  )
+
+  use_rextendr_template(
+    "cran/Makevars.win",
+    save_as = file.path("src", "Makevars.win"),
+    quiet = quiet,
+    overwrite = overwrite,
+    data = list(lib_name = lib_name)
+  )
+
+  # vendor directory should be ignored by git and R CMD build
+  if (!rlang::is_installed("usethis")) {
+    cli::cli_inform(
+      c(
+        "!" = "Add {.code ^src/rust/vendor$} to your {.file .Rbuildignore}",
+        "!" = "Add {.code ^src/rust/vendor$} to your {.file .gitignore}",
+        "i" = "Install {.pkg usethis} to have this done automatically."
+      )
+    )
+  } else {
+    # vendor folder will be large when expanded and should be ignored
+    usethis::use_build_ignore(
+      file.path("src", "rust", "vendor")
+    )
+
+    usethis::use_git_ignore(
+      file.path("src", "rust", "vendor")
+    )
+  }
+
+  invisible(NULL)
+}
+
+#' @export
+#' @name cran
+vendor_pkgs <- function(path = ".", quiet = FALSE, overwrite = NULL) {
+  local_quiet_cli(quiet)
+
+  # get path to rust folder
+  src_dir <- rprojroot::find_package_root_file(path, "src/rust")
+
+  # if `src/rust` does not exist error
+  if (!dir.exists(src_dir)) {
+    cli::cli_abort(
+      c("{.path src/rust} cannot be found", "i" = "Did you run {.fn use_extendr}?"),
+      class = "rextendr_error"
+    )
+  }
+
+  # if cargo.lock does not exist, cerate it using `cargo update`
+  cargo_lock_fp <- file.path(src_dir, "Cargo.lock")
+
+  if (!file.exists(cargo_lock_fp)) {
+    withr::with_dir(src_dir, {
+      update_res <- processx::run(
+        "cargo",
+        c(
+          "generate-lockfile",
+          "--manifest-path",
+          file.path(src_dir, "Cargo.toml")
+        ),
+        stderr_line_callback = function(x, proc) {
+          if (!grepl("To use vendored sources", x) && x != "") {
+            cli::cat_bullet(stringi::stri_trim_left(x))
+          }
+        }
+      )
+    })
+  }
+
+  if (update_res[["status"]] != 0) {
```

Comment:
I think this is a bug -- if `cargo_lock_fp` exists, `update_res` is never assigned before this line.

Reply:
ah yup! I fixed this locally but it got overwritten when i pulled! Thank you!